sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -17,16 +17,34 @@
 
 package org.apache.spark.sql.catalyst.plans
 
-import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeSet, Expression, VirtualColumn}
+import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.trees.TreeNode
 import org.apache.spark.sql.types.{DataType, StructType}
 
-abstract class QueryPlan[PlanType <: TreeNode[PlanType]] extends TreeNode[PlanType] {
+abstract class QueryPlan[PlanType <: TreeNode[PlanType]]
+  extends TreeNode[PlanType] with PredicateHelper {
   self: PlanType =>
 
   def output: Seq[Attribute]
 
   /**
+   * Extracts the relevant per-row constraints from a given child while removing those that
+   * don't apply anymore.
+   */
+  protected def getRelevantConstraints(child: QueryPlan[PlanType]): Set[Expression] = {
+    child.constraints.filter(_.references.subsetOf(outputSet))
+  }
+
+  /**
+   * A sequence of expressions that describes the data property of the output rows of this
+   * operator. For example, if the output of this operator is column `a`, an example `constraints`
+   * can be `Set(a > 10, a < 20)`.
+   */
+  lazy val constraints: Set[Expression] = validConstraints
+
+  protected def validConstraints: Set[Expression] = Set.empty
```

Comment:
Can we add doc here suggesting the child classes override this method along with the semantics (i.e. if we go with the suggestion above you are supposed to output any constraint that might be valid and it will canonicalized and filtered automatically).